designs/ami-selector-reqs.md

```diff
@@ -0,0 +1,244 @@
+selector itself.
+
+A conceptual implementation could look like:
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1
+kind: EC2NodeClass
+metadata:
+  name: default
+spec:
+  amiFamily: Custom
+  amiSelectorTerms:
+    # Standard AMI with no special requirements
+    - name: my-ami
+      owner: self
+
+    # An ML-focused AMI that requires a GPU instance
+    - ssmParameter: /my-org/amis/custom-ml-drivers
+      requirements:
+        - key: "karpenter.k8s.aws/instance-gpu-count"
+          operator: "Exists"
+```
+
+The specified requirements would be merged with any constraints that can be statically inferred from the AMI itself
+(e.g., architecture). For instance, given the above configuration, the `EC2NodeClass` status might resolve to:
+
+```yaml
+status:
+  amis:
+    - id: ami-0bc680bafd3cdf722
+      name: bottlerocket-aws-k8s-1.31-nvidia-x86_64-v1.42.0-5ed15786
+      requirements:
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+            - amd64
+        - key: karpenter.k8s.aws/instance-gpu-count
+          operator: Exists
+```
+
+In cases where a requirement (e.g., `kubernetes.io/arch`) is specified both in the user-defined requirements and
+inferred from the AMI itself, the inferred value should take precedence. This prevents misconfigurations, such
+as specifying `kubernetes.io/arch: amd64` for an arm64 AMI.
+
+> [!NOTE]
+> We chose the field name `requirements` to align with the existing `status.amis[*].requirements` field on the
+> `EC2NodeClass`. `nodeRequirements` was also considered for its explicitness.
+
+* **Pro**: The association between an AMI and its requirements is declarative and self-contained.
+* **Pro**: Avoids a breaking change to the `EC2NodeClass` schema (e.g., introducing a nested selector field).
+* **Con**: Slight semantic overload: `amiSelectorTerms` now includes both AMI selection and scheduling logic, which
+  may blur conceptual boundaries.
+    * On one hand, the term now includes both AMI selection logic and node scheduling constraints, which may blur the
+      conceptual boundary between "*what AMI to use*" and "*under what conditions it can be used*".
+    * On the other hand, while `requirements` is not a direct AMI selector, it meaningfully contributes to the AMI
+      selection process by constraining the set of instance types that are compatible with a given AMI. Karpenter
+      already selects AMIs based on instance compatibility and recency. These requirements act as an additional filter
+      on the AMI's applicability in a given scheduling context. Today, except alias-based terms, these constraints are
+      limited to what can be statically inferred from the AMI itself (e.g., architecture). Therefore, including
+      requirements as a field within an `amiSelectorTerm` is consistent with its purpose: to define the **AND**ed
+      conditions under which an AMI is valid for use.
```

Comment:
You've convinced me that this approach makes sense without significantly changing the semantic meaning of what an AMI selector term is. I was coming at it from the standpoint of our implementation, where AMI discovery and selection per instance-type are two distinct processes, but it really isn't from an end-user perspective. It would be good to get community consensus on this, but I'm bought in.